Socket parsing rework

keszybz opened the pull request.

poettering commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -456,25 +456,28 @@ static int manager_dns_stub_udp_fd(Manager *m) {
         return m->dns_stub_udp_fd = TAKE_FD(fd);
 }
 
-static int manager_dns_stub_udp_fd_extra(Manager *m, DNSStubListenerExtra *l) {
```
hmm, collides with @yuwata's #16951

poettering commented on src/basic/socket-util.h:
```diff
@@ -124,9 +124,14 @@ int fd_inc_rcvbuf(int fd, size_t n);
+      IFNAME_VALID_ALTERNATIVE = 1 << 0,
+      IFNAME_VALID_NUMERIC     = 1 << 1,
+      _IFNAME_VALID_ALL        = IFNAME_VALID_ALTERNATIVE | IFNAME_VALID_NUMERIC,
+};
```
why anonymous? i'd always use named enums for flags unless there's a good reason not to. it's nicer for debuggers and static analyzers

poettering commented on src/basic/socket-util.c:
```diff
@@ -399,19 +399,25 @@ int sockaddr_pretty(
+                        char a[INET6_ADDRSTRLEN], ifname[IF_NAMESIZE + 1];
 
                         inet_ntop(AF_INET6, &sa->in6.sin6_addr, a, sizeof(a));
+                        if (sa->in6.sin6_scope_id != 0)
```
i wonder if we should conditionalize this further: check if this is from the ipv6 link-local range, and if not suppress it to, since it's ignored by the kernel then iirc

poettering replied:
it's even documented in ipv6(7)

keszybz replied:
I'd say that the printer should not do the suppression. It should faithfully report what is in the struct. (It's just confusing if the printer is too smart.) IMHO, instead we could detect such misuse in the parser, and warn and ignore. This would also be much more useful for users.

poettering commented:
I guess this needs to be aligned with @yuwata's work. Maybe merge #16951 first and then rebase this? or the other way round? you guys figure that out!

keszybz commented:
> I guess this needs to be aligned with @yuwata's work. Maybe merge #16951 first and then rebase this? or the other way round? you guys figure that out!

Yeah, I already wrote in #16951 that it should be merged first.

flo-wer commented on src/shared/socket-netlink.c:
```diff
@@ -372,10 +332,14 @@ int in_addr_port_ifindex_name_from_string_auto(
-
-        /* if ret_port is NULL, then strings with port cannot be specified.
-         * Also, if ret_server_name is NULL, then server_name cannot be specified. */
+         * [2001:4860:4860::8888]:53%eth0#example.com
```
Parsing of the ifindex in this way is ambiguous and incomplete. [RFC4007 11.7](https://tools.ietf.org/html/rfc4007#section-11.7) specifies how the zone id aka ifindex is used together with other delimiter characters. Though the RFC admits some ambiguity and "does not specify how the format for non-global addresses should be combined with the preferred format for literal IPv6 addresses", it does give a example `fe80::%2/64` and specifies that "it is important to place the zone index portion before the prefix length".

From this I conclude that the preferred way for this type of address is `[fe80::42%eth0]:53#example.com`, but only if the address is link-local and the ifindex is the "zone id" for this address. On the other hand, if the ifindex is used only to bind the socket to this interface, the preferred way is  the existing `[2001:4860:4860::8888]:53%eth0#example.com`.

So I suggest to separate these two types of ifindex, as they have different meaning in different contexts. Though in all cases both would have the same value.

keszybz replied:
I'd be against having two syntaxes for specifying the index. We would never need both at the same time, so it seems fine to just use the second one for both purposes.

poettering commented:
ah, christ, i just realized that much of the work i did in #16985 is already done in this PR...

poettering commented on src/resolve/resolved-bus.c:
```diff
@@ -1600,7 +1600,7 @@ static int bus_property_get_ntas(
 }
 
-static BUS_DEFINE_PROPERTY_GET_ENUM(bus_property_get_dns_stub_listener_mode, dns_stub_listener_mode, DnsStubListenerMode);
+static BUS_DEFINE_PROPERTY_GET_ENUM(bus_property_get_dns_stub_listener_mode, dns_stub_listener_mode, DNSStubListenerMode);
```
hmm, we use the lowercase spelling for pretty much *all* structs/enums so far, DNSStubListenerExtra is the one outlier. That's why in #16985 i wen the other way, and fixes that one to use lower case spelling

poettering replied:
i.e. compare DnsPacket, DnsResourceRecord, DnsStream …

keszybz replied:
Hmm, I asked @ssahani to use uppercase in another PR. I guess consistent "Dns" spelling is OK too.

poettering commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -502,19 +502,21 @@ static int set_dns_stub_common_socket_options(int fd, int family) {
 }
 
-static int manager_dns_stub_udp_fd(Manager *m) {
+static int manager_dns_stub_fd(Manager *m, bool udp) {
```
why not pass "int type" here, that is SOCK_DRGAM or SOCK_STREAM. Would be much nicer I think

poettering commented on src/basic/socket-util.c:
```diff
@@ -399,19 +399,25 @@ int sockaddr_pretty(
                         } else {
-                                p = strdup(a);
+                                p = strjoin(a,
+                                            sa->in6.sin6_scope_id != 0 ? "%s" : NULL,
```
s too much

poettering commented on src/basic/socket-util.c:
```diff
@@ -399,19 +399,25 @@ int sockaddr_pretty(
-                                p = strdup(a);
+                                p = strjoin(a,
+                                            sa->in6.sin6_scope_id != 0 ? "%s" : NULL,
+                                            sa->in6.sin6_scope_id != 0 ? ifname : NULL);
```
i'd just write this as:

```c
p = sa->in6.sin6_scope_id != 0 ? strjoin(a, "%", ifname) : strdup(a);
```

i think that is much prettier

poettering commented on src/shared/socket-netlink.c:
```diff
@@ -62,44 +62,42 @@ int socket_address_parse(SocketAddress *a, const char *s) {
+                        .size = offsetof(struct sockaddr_un, sun_path) + l + 1,
+                };
                 memcpy(a->sockaddr.un.sun_path, s, l);
-                a->size = offsetof(struct sockaddr_un, sun_path) + l + 1;
```
hmm, we should probably rewrite this to just use sockaddr_un_set_path()

poettering commented:
> ah, christ, i just realized that much of the work i did in #16985 is already done in this PR...

Actually, I take that back, it's not as overlapping as I thought. But probably one will need a rebase if the other is merged

keszybz commented:
Updated. I dropped the commit to rename DnsStubListenerMode. The commits to merge the stub functions are heavily reworked, fsee the commit comments. The suggestion to pass `type` is implemented. I also added two new commits to use sockaddr_un_set_path() as suggested.

keszybz commented:
centos ci:
```
02:05:05 ======================================================================
02:05:05 FAIL: test_lldp (__main__.NetworkdLLDPTests)
02:05:05 ----------------------------------------------------------------------
02:05:05 Traceback (most recent call last):
02:05:05   File "test/test-network/systemd-networkd-tests.py", line 3081, in test_lldp
02:05:05     self.assertRegex(output, 'veth-peer')
02:05:05 AssertionError: Regex didn't match: 'veth-peer' not found in 'LINK             CHASSIS ID        SYSTEM NAME      CAPS        PORT ID           PORT DESCRIPTION\n\n0 neighbors listed.'
02:05:05 
02:05:05 ----------------------------------------------------------------------
```
Prolly unrelated.

poettering commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -632,50 +638,67 @@ static int manager_dns_stub_udp_fd_extra(Manager *m, DnsStubListenerExtra *l) {
+        else
+                r = setsockopt_int(fd, IPPROTO_IPV6, IPV6_FREEBIND, true);
+        if (r < 0)
+                goto fail;
```
this would be much prettier if based on #17005's socket_set_freebind() helper!

keszybz replied:
Interesting. One person touches a piece of code and then it cascades into a series of lengthy prs. I guess this code was ripe for some cleanups.

poettering commented:
I will rebase #17005 on top of this, and update the code I referenced above.
